### PR TITLE
[java] Add support for basic method type inference for strict invocation

### PR DESCRIPTION
**Pull request by WinterGrascph**

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTExpression.java [hunk 1]
```diff
@@ -20,4 +20,30 @@ public ASTExpression(JavaParser p, int id) {
     public Object jjtAccept(JavaParserVisitor visitor, Object data) {
         return visitor.visit(this, data);
     }
+
+    public boolean isStandAlonePrimitive() {
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTExpression.java]: I'm a little confused by this method.... It is named as to recognized primitives, but will actually recognize literals, that are not necessarily primitives (ie: a String literal). Also, the last check `literal.jjtGetNumChildren() != 0` means we are not considering the primitive booleans (`true` and `false`) as stand alone primitives.

`null` would also not be recognized as a primitive, but that one is probably in line with the method name.

I'm not sure how much of the name should be changed vs the implementation, but definitely needs some rework

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typeinference/TypeInferenceResolver.java [hunk 2]
```diff
@@ -31,6 +32,38 @@ private TypeInferenceResolver() {
 
     }
 
+    public static List<JavaTypeDefinition> inferTypes(List<Constraint> constraints, List<Bound> bounds,
+                                                      List<Variable> variables) {
+
+        List<Bound> newBounds = new ArrayList<>();
+        while (!constraints.isEmpty()) {
+            List<BoundOrConstraint> reduceResult = constraints.get(constraints.size() - 1).reduce();
+            constraints.remove(constraints.size() - 1);
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typeinference/TypeInferenceResolver.java]: these 2 lines could be condensed. `remove` returns the removed element, so you could directly write: `List<BoundOrConstraint> reduceResult = constraints.remove(constraints.size() - 1).reduce();`

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTExpression.java (hunk 1 above)
Comment (WinterGrascph) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTExpression.java]: It was supposed to recognize primitives, the name is not missleading, the implementation is faulty.

Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTExpression.java]: Amazing!

**Pull request merged**